Documenting Migrators and Migrations. (#1801)
conda-forge/conda-forge.github.io: src/maintainer/knowledge_base.rst

So make sure that you **tag** your package in such a way that the package name
that conda-build spits out will sort the package uploaded with an ``rc`` label
higher than the package uploaded with the ``dev`` label.

Migrators and Migrations
========================

> When any changes are made in the global pinnings of a package, then the entire stack of the packages that depend on that package would need to be updated and rebuilt.

Reviewer: Not all packages that depend on a pinned package need to be rebuilt. We only need to rebuilt those that impose an ABI dependency (usually via linking / dynamic libraries, but sometimes also header libraries like pybind11). I am not sure of what the right terminology is here, though, but the rebuild only concerns packages which depend on pinned packages in their `host` section.

```suggestion
When any changes are made in the global pinnings of a package, then the entire stack of the packages that need that package on their `host` section would need to be updated and rebuilt.
```

> Doing it manually could be quite tedious, and that's where migrations come to help. Migrations are used to change how different feedstocks/packages operate and are an integral part of ``regro-cf-autotick-bot``'s duties.

Reviewer: ```suggestion
Doing it manually could be quite tedious, and that's where migrations come to help. Migrations automate the process of submitting changes to a feedstock and are an integral part of ``regro-cf-autotick-bot``'s duties.
```

Reviewer 2: I'd also suggest using present tense where possible.

```suggestion
Doing it manually can be quite tedious, and that's where migrations come to help. Migrations are used to change how different feedstocks/packages operate and are an integral part of the ``regro-cf-autotick-bot``'s duties.
```

> To generate these migrations, you use migrators. Migrators automatically create pull requests for the affected packages in conda-forge. 

Reviewer: There are several types of migrations. Most of the time we only need an updated pin, which can be triggered via migration files. However there are others. I'd mention the `cf-scripts` docs: https://regro.github.io/cf-scripts/migrators.html too

> To propose a migration in one or more pins, a PR will be issued into the pinning feedstock using a yaml file expressing the changes to the global pinning file in the migrations folder. 
> Once the PR is squashed and merged, the bot will take the yaml file to calculate which packages need to be rebuilt and start issuing PRs with the new pinning as a local pinning file to the feedstocks.
> 
> Usually, the bot will generate these migrations automatically. However, when a pin is first made or added, one may need to be added by hand. To do this, you can follow the steps mentioned in `Updating package pins <https://conda-forge.org/docs/maintainer/pinning_deps.html#updating-package-pins>`__.

Reviewer 2: Same here with present tense.

```suggestion
To propose a migration in one or more pins, a PR is issued into the pinning feedstock using a yaml file expressing the changes to the global pinning file in the migrations folder. 
Once the PR is squashed and merged, the bot takes the yaml file to calculate which packages need to be rebuilt and starts issuing PRs with the new pinning as a local pinning file to the feedstocks.

Usually, the bot generates these migrations automatically. However, when a pin is first made or added, one may need to be added by hand. To do this, follow the steps mentioned in `Updating package pins <https://conda-forge.org/docs/maintainer/pinning_deps.html#updating-package-pins>`__.
```

The way migrations proceed are: 

  1. You make a PR into the ``migrations`` folder in the `conda-forge-pinning-feedstock <https://github.com/conda-forge/conda-forge-pinning-feedstock>`__ with a new YAML file representing the migration.
  2. Once the PR is merged, the bot picks it up, builds a migrator object, and begins the migration process.
  3. A migration PR is issued for a node only if:
      - The node depends on the changed pinnings.

>       - It has no dependencies that depend on the new pinnings and have not been migrated.

Reviewer 2: Does the "it" in that second bullet point refer to the node or something else?

Author: @kathatherine, it refers to node. Should I replace "It" with "the node"?

Reviewer 2: @ssurbhi560 Yes, please! Would help with clarity.

  4. Process 3 continues until the migration is sufficiently complete and the change is applied to the global pinning file via a PR.

Sometimes, you might get a migration PR for your package that you don't want to merge. In that case, it is advisable to put that PR in draft status rather than closing it. 

> If you will close the PR, it would make that bot consider PR as merged. And that can cause inconsistencies with downstream feedstocks.

Reviewer: I would use stronger wording here. We kindly ask people do NOT close PRs if they are unwilling to work on it. Reasoning:

- The bot will assume another PR implementing the migration was merged instead, letting the migration continue iterating on the graph; however downstream dependents will fail because the parent (the one we closed the PR of) didn't really get built.
- Somebody might pass by and help if they see an open PR

> If you will close the PR, it would make that bot consider PR as merged. And that can cause inconsistencies with downstream feedstocks.

Reviewer 2: If you keep any of the language in this sentence after going through Jaime's comment above, here's some suggestions for the grammar.

```suggestion
If you close the PR, it makes the bot consider that PR as merged and that can cause inconsistencies with downstream feedstocks.
```

So make sure that you **tag** your package in such a way that the package name
that conda-build spits out will sort the package uploaded with an ``rc`` label
higher than the package uploaded with the ``dev`` label.

Migrators and Migrations
========================

When any changes are made in the global pinnings of a package, then the entire stack of the packages that need that package on their ``host`` section would need to be updated and rebuilt.
Doing it manually can be quite tedious, and that's where migrations come to help. Migrations automate the process of submitting changes to a feedstock and are an integral part of the ``regro-cf-autotick-bot``'s duties.

There are several kinds of migrations, to know about different types of migration you can read `Making Migrators <https://regro.github.io/cf-scripts/migrators.html>`__. To generate these migrations, you use migrators. Migrators automatically create pull requests for the affected packages in conda-forge.
To propose a migration in one or more pins, a PR is issued into the pinning feedstock using a yaml file expressing the changes to the global pinning file in the migrations folder.

> Once the PR is squashed and merged, the bot takes the yaml file to calculate which packages need to be rebuilt and starts issuing PRs with the new pinning as a local pinning file to the feedstocks.

Reviewer 2: ```suggestion
Once the PR is squashed and merged, the migrator bot uses the yaml file to calculate which packages need to be rebuilt and starts issuing PRs with the new pinning as a local pinning file to the feedstocks.
```

Usually, the bot generates these migrations automatically. However, when a pin is first made or added, one may need to be added by hand. To do this, you can follow the steps mentioned in `Updating package pins <https://conda-forge.org/docs/maintainer/pinning_deps.html#updating-package-pins>`__.

The way migrations proceed are: 

  1. You make a PR into the ``migrations`` folder in the `conda-forge-pinning-feedstock <https://github.com/conda-forge/conda-forge-pinning-feedstock>`__ with a new YAML file representing the migration.
  2. Once the PR is merged, the bot picks it up, builds a migrator object, and begins the migration process.
  3. A migration PR is issued for a node only if:

    - The node depends on the changed pinnings.
    - The node has no dependencies that depend on the new pinnings and have not been migrated.

  4. Process 3 continues until the migration is sufficiently complete and the change is applied to the global pinning file via a PR.

Sometimes, you might get a migration PR for your package that you don’t want to merge. In that case, you should put that PR in draft status but should never close it.

> If you close the PR, it makes the bot consider that another PR implementing the migration is merged instead, letting the migration continue iterating on the graph; however, the downstream dependents fails because the parent (the one we closed the PR of) didn’t really get built.

Reviewer 2: ```suggestion
If you close the PR, it makes the bot think that another PR implementing the migration is merged instead, letting the migration continue iterating on the graph; however, the downstream dependents fail because the parent (the one we closed the PR of) didn’t really get built.
```

Once the PR is merged, the migration bot goes through the list of feedstocks in ``arch_rebuild.txt`` and opens a migration PR for any new feedstocks and their dependencies, enabling the aarch64/ppc64le builds.

.. _migrations_and_migrators:

Migrators and Migrations
========================

When any changes are made in the global pinnings of a package, then the entire stack of the packages that need that package on their ``host`` section would need to be updated and rebuilt.
Doing it manually can be quite tedious, and that's where migrations come to help. Migrations automate the process of submitting changes to a feedstock and are an integral part of the ``regro-cf-autotick-bot``'s duties.

There are several kinds of migrations, which you can read about in `Making Migrators <https://regro.github.io/cf-scripts/migrators.html>`__. To generate these migrations, you use migrators, which are bots that automatically create pull requests for the affected packages in conda-forge.
To propose a migration in one or more pins, the migrator issues a PR into the pinning feedstock using a yaml file expressing the changes to the global pinning file in the migrations folder.

> Once the PR is squashed and merged, the migrator bot uses the yaml file to calculate which packages need to be rebuilt and starts issuing PRs with the new pinning as a local pinning file to the feedstocks.

Reviewer 3: Why do we have to mention "squash" here? Please don't... It's a very bad practice against tracking changes, sometimes a migration PR needs additional work by the maintainers. I understand there are "purists" out there who would defend the cleanness of the commit history, but really it's unnecessary to advocate.
```suggestion
Once the PR is merged, the migrator bot uses the yaml file to calculate which packages need to be rebuilt and starts issuing PRs with the new pinning as a local pinning file to the feedstocks.
```

Reviewer 3: (This sentence is also a bit weird to me, IIRC we build the whole dependency graph first, and then the bot walks through the graph to migrate all nodes one by one. @beckermr can chime in here.)

Author: Hi @leofang! Thank you so much for reviewing the PR. 
Rephrasing it to - "Once the PR is merged, the dependency graph is built. After that, the bot walks through the graph and migrates all the nodes (feedstocks) one by one and issues PRs for those feedstocks." would this be correct? 
@kathatherine, can you also please take a look and let me know if something needs to be changed in this?

Reviewer 2: @ssurbhi560 Let's avoid two "and"s in the same list.

"Once the PR is merged, the dependency graph is built. After that, the bot walks through the graph, migrates all the nodes (feedstocks) one by one, and issues PRs for those feedstocks."

Would also love @beckermr's check of this.

Reviewer 4: Yes @leofang is correct. We issue PRs in topological order in the graph and downstream packages have to wait for their upstream parents before they get a PR.

Usually, the bot generates these migrations automatically. However, when a pin is first made or added, one may need to be added by hand. To do this, you can follow the steps mentioned in `Updating package pins <https://conda-forge.org/docs/maintainer/pinning_deps.html#updating-package-pins>`__.

The way migrations proceed are: 

>   1. You make a PR into the ``migrations`` folder in the `conda-forge-pinning-feedstock <https://github.com/conda-forge/conda-forge-pinning-feedstock>`__ with a new YAML file representing the migration.

Reviewer 3: I noticed we use both lowercase "yaml" and uppercase "YAML" in this section. Can we unify them?

>   2. Once the PR is merged, the bot picks it up, builds a migrator object, and begins the migration process.

Reviewer 3: ```suggestion
  2. Once the PR is merged, the bot picks it up, builds a migrator graph, and begins the migration process.
```

>   3. A migration PR is issued for a node only if:

Reviewer 3: Up to this point we haven't mentioned anything about graphs or nodes, so we should be clear what we're referring to:
```suggestion
  3. A migration PR is issued for a node (a feedstock) only if:
```

    - The node depends on the changed pinnings.
    - The node has no dependencies that depend on the new pinnings and have not been migrated.

>   4. Process 3 continues until the migration is sufficiently complete and the change is applied to the global pinning file via a PR.

Reviewer 3: ```suggestion
  4. Process 3 continues until the migration is complete and the change is applied to the global pinning file via a final PR. After this step, we say this migration is closed out.
```

Sometimes, you might get a migration PR for your package that you don’t want to merge. In that case, you should put that PR in draft status but should never close it.

> If you close the PR, it makes the bot think that another PR implementing the migration is merged instead, letting the migration continue iterating on the graph; however, the downstream dependents fail because the parent (the one we closed the PR of) didn’t really get built.

Reviewer 3: Note: here we talk about the graph.

Reviewer 3: nit:
```suggestion
If you close the PR, it makes the bot think that another PR implementing the migration is merged instead, letting the migration continue iterating on the graph; however, the downstream dependents fail because the parent (the one we closed the PR of) didn’t really get rebuilt.
```
